lib/Sema/DerivedConformanceCodable.cpp

```diff
@@ -0,0 +1,949 @@
+
+  // Bind Keyed*Container to Keyed*Container<KeyType>
+  Type boundType[1] = { keyType };
+  auto containerType = BoundGenericType::get(keyedContainerDecl, Type(),
+                                             C.AllocateCopy(boundType));
+
+  // let container : Keyed*Container<KeyType>
+  auto containerDecl = new (C) VarDecl(/*IsStatic=*/false, /*IsLet=*/true,
+                                       /*IsCaptureList=*/false, SourceLoc(),
+                                       C.Id_container, containerType, DC);
+  containerDecl->setImplicit();
+  containerDecl->setInterfaceType(containerType);
+  return containerDecl;
+}
+
+static CallExpr*
+createContainerKeyedByCall(ASTContext &C, DeclContext *DC, Expr *base,
+                           Type returnType, NominalTypeDecl *param) {
+  // (keyedBy:)
+  auto keyedByDecl = new (C) ParamDecl(/*IsLet=*/true, SourceLoc(),
+                                       SourceLoc(), C.Id_keyedBy,
+                                       SourceLoc(), C.Id_keyedBy,
+                                       returnType, DC);
+  keyedByDecl->setImplicit();
+  keyedByDecl->setInterfaceType(returnType);
+
+  // container(keyedBy:) method name
+  auto paramList = ParameterList::createWithoutLoc(keyedByDecl);
+  DeclName callName(C, C.Id_container, paramList);
+
+  // base.container(keyedBy:) expr
+  auto unboundCall = new (C) UnresolvedDotExpr(base, SourceLoc(), callName,
+                                               DeclNameLoc(),
+                                               /*Implicit=*/true);
+
+  // CodingKeys.self expr
+  auto codingKeysExpr = new (C) DeclRefExpr(ConcreteDeclRef(param),
+                                            DeclNameLoc(), /*Implicit=*/true);
+  auto codingKeysMetaTypeExpr = new (C) DotSelfExpr(codingKeysExpr, SourceLoc(),
+                                                    SourceLoc());
+
+  // Full bound base.container(keyedBy: CodingKeys.self) call
+  Expr* args[1] = { codingKeysMetaTypeExpr };
+  Identifier argLabels[1] = { C.Id_keyedBy };
+  return CallExpr::createImplicit(C, unboundCall, C.AllocateCopy(args),
+                                  C.AllocateCopy(argLabels));
+}
+
+static void deriveBodyCodable_encode(AbstractFunctionDecl *encodeDecl) {
+  // struct Foo : Codable {
+  //   var x: Int
+  //   var y: String
+  //
+  //   // Already derived by this point if possible.
+  //   @derived enum CodingKeys : CodingKey {
+  //     case x
+  //     case y
+  //   }
+  //
+  //   @derived func encode(to encoder: Encoder) throws {
+  //     let container = encoder.container(keyedBy: CodingKeys.self)
+  //     try container.encode(x, forKey: .x)
+  //     try container.encode(y, forKey: .y)
+  //   }
+  // }
+
+  // The enclosing type decl.
+  auto typeDecl = cast<NominalTypeDecl>(encodeDecl->getDeclContext());
+
+  auto funcDC = cast<DeclContext>(encodeDecl);
+  ASTContext &C = funcDC->getASTContext();
+  auto &tc = *(TypeChecker *)C.getLazyResolver();
```

Comment:
If you're just doing a conformance check and not trying to record a dependency or emit a diagnostic, you can call ModuleDecl::lookupConformance() instead of TypeChecker::conformsToProtocol().